Compose files templating improvements

tgaugry opened the pull request:
This PR add a few things related to compose file templating (see #19).
This adds :
- A way to have common values for all stacks, instead of specifying the same variable over and over in each stack values_file. These can be specified either in the "global_values" field in config.yaml, or in a file named global.yaml (same behavior as the "stacks" variable. They can be overriden in stack values_file, if needs be.

- You can now use [sprig](https://masterminds.github.io/sprig/) functions in compose.
- You can use go templates defined in the "template" folder.
- The web UI now shows original and generated files (if available)
<img width="616" height="200" alt="image" src="https://github.com/user-attachments/assets/b092a9d1-fcde-4a7a-9bf1-c199ef461f37" />

Commit:
Add global values file option

Commit:
Move global values to global.yaml file

Global values can also be written directly in config.yaml
This makes behavior uniform across repos and stacks configuration

Commit:
feat: Compose files can include templated parts

Commit:
Allow sprig in templates

Commit:
Add links in the iterface for templates and compose files

sanzoghenzo reviewed (changes requested):
OK, I think I understand your changes.

I don't use the WEB UI, so to me it is a bit much having to manage a "cache" directory (doesn't seem like a cache since it is recreated at  each update without checking if anything changed - but that's just semantic and I get what you wanted to do 😉 ) and perform I/O operations just to be able to download the rendered compose file.

I'll leave to @m-adawi  to decide, but couldn't this be turned into an on-demand API?

sanzoghenzo commented:
> * Problem is that people will have no way to check what was generated

I was thinking about this just today, I would like to test a templated stack locally without the need to spin up a swarmCD instance...
It would be great to have a cli tool that does only that!
This would mean extracting the core logic to a separate module that can then be called by both swarmcd and the tool (that could be a sub-command of the current program if a separate program is overkill).

I'm not saying that you need to do that 😅 , it's just a thought... I'll make this into an issue ASAP!

tgaugry commented:
I tinkered a bit yesterday, and you were right, solution 3 is not that complex - I was tricked by the function export rules from go (starting with an upper case letter vs not).

> This would mean extracting the core logic to a separate module that can then be called by both swarmcd and the tool (that could be a sub-command of the current program if a separate program is overkill).

Honestly, with solution 3 I feel like a good chunk of the work is done for that. I'll go that route, and I'll see if I can find the time for #127

Commit:
Add template-gen executable + refactor + readme

Commit:
Add link to doc

sanzoghenzo commented:
Looks good to me! I still didn't test it, though.

@m-adawi have you any thought about this PR? should I proceed with the merge or do you want to try it first?

tgaugry commented:
Didn't do any real changes, just rebased.

I have not seen any problems for the last 3ish months, so I guess this is fine ? (The most famous last words "works on my pc"...)

@sanzoghenzo Is there any blocker to tackle before this can get merged ?

sanzoghenzo commented on web/controllers.go:
```diff
@@ -10,17 +10,46 @@ import (
+	swarmStack, err := swarmcd.GetSwarmStack(stackName)
+	if err != nil {
+		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+	}
```
I believe a `return` is missing here to exit early from the function

sanzoghenzo commented on web/controllers.go:
```diff
@@ -10,17 +10,46 @@ import (
+	stackBytes, err := swarmStack.ReadStack()
+	if err != nil {
+		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+	}
```
add `return`

sanzoghenzo commented on web/controllers.go:
```diff
@@ -10,17 +10,46 @@ import (
+	swarmStack, err := swarmcd.GetSwarmStack(stackName)
+	if err != nil {
+		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+	}
```
add `return`

sanzoghenzo commented on web/controllers.go:
```diff
@@ -10,17 +10,46 @@ import (
+	stackBytes, err := swarmStack.GenerateStack()
+	if err != nil {
+		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+	}
```
add `return`

sanzoghenzo commented on util/values_map.go:
```diff
@@ -0,0 +1,17 @@
+                return nil, fmt.Errorf("could not read %s values file: %w", source, err)
+        }
+        var valuesMap map[string]any
+        yaml.Unmarshal(valuesBytes, &valuesMap)
```
Unmarshal can return an error, it's better to handle it

sanzoghenzo commented on cmd/template-gen/main.go:
```diff
@@ -0,0 +1,81 @@
+		fmt.Println(string(stackBytes))
+
+	} else {
+		err = os.WriteFile(outputFile, stackBytes, 0666)
```
maybe use mode 600?

sanzoghenzo commented on swarmcd/stack.go:
```diff
@@ -41,27 +59,19 @@ func (swarmStack *swarmStack) updateStack() (revision string, err error) {
 	if err != nil {
 		return
 	}
+	stackStatus[swarmStack.name].Templated = swarmStack.templated
```
AI suggests to check for key existence... I don't think is needed, but it doesn't hurt. Especially if/when we go ahead with the dynamic configuration discovery

tgaugry replied:
Make no real sense as they are created at the same time (and i don't see a use case to decorrelate them). I'll add it anyway.

sanzoghenzo commented on swarmcd/stack.go:
```diff
@@ -94,35 +104,85 @@ func (swarmStack *swarmStack) updateStack() (revision string, err error) {
+		log.Debug("Loading template folder...")
+		
+		pattern := path.Join(swarmStack.templateFolder, "*.tmpl")
+		filenames, err := filepath.Glob(pattern)
+		if err == nil {
+			if len(filenames) != 0 {
+				_, err = templ.ParseFiles(filenames...)
+			} else {
+				log.Debug("Skipping, folder empty", "folder", swarmStack.templateFolder)
+			}
+		}
+		if err != nil {
+			log.Error("Could not parse templates, trying to generate stack without them.", "error", err)
+		}
```
From the docs:

> Glob returns the names of all files matching pattern or nil if there is no matching file. 
> The only possible returned error is `ErrBadPattern`, when pattern is malformed.

so the right check for the non-empty folder is `if filenames != nil`, and we might be more explicit about the bad pattern (well, you already added the path existence check, so bad patterns shouldn't creep up...)

tgaugry replied:
If there is a bad pattern, then l152 is false and it is caught l159. I don't see how they would creep up ?

sanzoghenzo replied:
What I was saying is that right now the bad pattern error is handled with the same message as if it was a parsing error, but I don't believe they are the same thing, and a bad pattern should stop the execution.
For "errors creeping up" I was referring to the fact that the pattern is already checked beforehand (you already made sure the template folder exists), so there shouldn't be any chance the bad pattern error will come up.. unless there some non obvous way that the pattern is modified.
Long story short: leave it like this :wink:

sanzoghenzo commented on README.md:
```diff
@@ -136,6 +136,101 @@ Please note that:
+
+### Defining global variables
+
+Variables can be defined for all stack in a `global_values.yaml` file, or directly in the `config.yaml` file using the `global_values` field.
```
```suggestion
Variables can be defined for all stacks in a `global_values.yaml` file, or directly in the `config.yaml` file using the `global_values` field.
```

sanzoghenzo commented on util/config.go:
```diff
@@ -52,13 +54,24 @@ func LoadConfigs() (err error) {
 	}
+	if Configs.GlobalValues == nil {
+		if _, errStat := os.Stat("global_values.yaml"); errStat == nil {
+			Configs.GlobalValues, err = ParseValuesFile("global_values.yaml", "global")
```
should we make the filename configurable? also, is it possible to use viper for consistency here (and drop values_map.go)?

tgaugry replied:
The function inside isn´t actually new, just a more generic version of what is used for the stack specific value files (embedded in `renderComposeTemplate`).
I'll move it back to `stacks.go`

sanzoghenzo commented on swarmcd/stack.go:
```diff
@@ -22,9 +25,21 @@ type swarmStack struct {
+
+	_, err := os.Stat(templateFolder)
+	if err != nil {
+		templateFolder = ""
```
this silently resets the folder even on permission errors, should we handle them explicitly?

tgaugry replied:
Added an error log - I let it continue, so the non templated stacks are not blocked by it.

sanzoghenzo commented on Dockerfile:
```diff
@@ -16,14 +16,17 @@ COPY cmd/ cmd/
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=${TARGETARCH} go build -o /swarm-cd ./cmd/
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=${TARGETARCH} go build -o /swarm-cd ./cmd/swarm-cd
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=${TARGETARCH} go build -o /template-gen ./cmd/template-gen
+RUN go test ./swarmcd/
```
remove the test from the dockerfile, we handle it in a separate CI job now

sanzoghenzo reviewed (changes requested):
AI was smarter than me in spotting things, here's the review.

It would be wonderful if you could also add some tests to cover the new features.

Commit:
Review changes + Fix config file paths

tgaugry commented:
It made no sense to me to have the global values file configurable, while `stacks` and `repos` aren't.
So I made all 3 configurable :smile: 

In the process, I encountered a bug: stacks and repos files were path relative to the working directory. I changed it to be relative to the main config file.

I have resolved everything, except one that I don't think I quite understand.

As for the tests, I have made those for the UI. I don't see any other than the ones in `swarmcd/stack_test.go`, and they don't cover the base functions I'm building on. I would have to basically write the tests for the whole project, and this is an endeavor quite beyond the scope of this PR. I however would be happy to contribute in another PR to add those when the base tests are covered.

sanzoghenzo commented:
I'm sorry I wa sloppy on the review, it was too late for me :sweat_smile:

We don’t need the configurable file names after all, and the config folder path is already taken care in one of my recent PRs awaiting approval.

I was deranged by the hardcoded filename twice, wile viper was hiding this to my tired eyes...

Can you restore these edits, please?

tgaugry commented:
EDIT: Yeah, no, disregard that. I'll make the template generator use explicit paths only, with a warn if a config with no embedded globals is provided.

You mean https://github.com/m-adawi/swarm-cd/pull/143 ?
Then the bug is still there. `"."` is correct for main config file, it is not for the other ones.
Here an example:
```
├── ex1
│   ├── config.yaml
│   ├── repos.yaml
│   └── stacks.yaml
├── ex2
│   ├── repos.yaml
│   ├── stacks.yaml
│   └── sub
│       └── config.yaml
└── ex3
    └── sub
        ├── config.yaml
        ├── repos.yaml
        └── stacks.yaml
```
When the working directory is ex1, it works with both versions.
When the working directory is ex2, using the config in `sub/config.yaml` works with the previous version (but it is quite strange). 
When the working directory is ex3, previous version does not work when using config in `sub/config.yaml`.

Main executable only load from working directory, so this is fine. The template generator, however, can load from an arbitrary path, so it will stumble.

Commit:
Remove _path vars

tgaugry commented:
> I'm sorry I wa sloppy on the review, it was too late for me 😅

It's ok - I don't think I was in a much better state TBH.

I have partially restored the edits to those files; I have kept the viper change.

sanzoghenzo commented on README.md:
````diff
@@ -136,6 +136,101 @@ Please note that:
+
+[Sprig](https://masterminds.github.io/sprig/) functions can be used in the compose files.
+
+```yaml
````
Add the stack example with the `value_file` field to be as explicit as possible

sanzoghenzo commented on README.md:
````diff
@@ -136,6 +136,101 @@ Please note that:
+---
+var_b: overriden
+```
+
````
add the stack example also here, without it it isn't unclear why the overridden value is used

sanzoghenzo commented on README.md:
````diff
@@ -136,6 +136,101 @@ Please note that:
+```yaml
+# value_file.yaml
+---
+var_b: overriden
````
```suggestion
var_b: overridden
```

sanzoghenzo commented on README.md:
```diff
@@ -136,6 +136,101 @@ Please note that:
+    image: foo
+    environment:
+      MY_VAR_A: "{{ .Values.var_a }}" # == fromglobal
+      MY_VAR_B: "{{ .Values.var_b }}" # == overriden
```
```suggestion
      MY_VAR_B: "{{ .Values.var_b }}" # == overridden
```

sanzoghenzo reviewed (changes requested):
Thanks for the work!

I added some comments for the documentation.

I also kindly ask you to run `go fmt` to fix some formatting.

> I don't see any other than the ones in swarmcd/stack_test.go, and they don't cover the base functions I'm building on. I would have to basically write the tests for the whole project, and this is an endeavor quite beyond the scope of this PR. 

Which base functions are you talking about? You basically rewrote the entire template generation feature, and that depends only on the swarmStack object and some files, so you can just create them (object + temporary files) and run GenerateStack againts it, then check that the output is what we expect.

Commit:
Readme + go fmt

Commit:
Add test for stack generation

tgaugry commented:
I have clarified the examples.

For the tests, I think I failed to get my point across, but that's ok, since we found a common ground on the scope in the end. They should cover most use cases.
<img width="954" height="557" alt="image" src="https://github.com/user-attachments/assets/4c405017-5314-479b-bb9c-dfaf16ddca0a" />

sanzoghenzo commented on cmd/main.go:
this file was removed instead of being moved to the swarm-cd subfolder as the new Dockerfile expects

tgaugry replied:
ah indeed, it get matched in .gitignore.

sanzoghenzo commented on swarmcd/stack.go:
```diff
@@ -94,35 +117,84 @@ func (swarmStack *swarmStack) updateStack() (revision string, err error) {
-	err = templ.Execute(&stackContents, map[string]map[string]any{"Values": valuesMap})
+	err = templ.Execute(&stackContents, map[string]map[string]any{"Values": mergedValuesMap})
 	if err != nil {
 		return nil, fmt.Errorf("error rending %s stack compose template: %w", swarmStack.name, err)
```
I know this is not your fault, but since we're at it, let's fix this typo 😉 

```suggestion
		return nil, fmt.Errorf("error rendering %s stack compose template: %w", swarmStack.name, err)
```

sanzoghenzo commented on web/controllers.go:
```diff
@@ -10,17 +10,50 @@ import (
+		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+	ctx.String(200, string(stackBytes))
```
```suggestion
	ctx.String(http.StatusOK, string(stackBytes))
```

sanzoghenzo commented on web/controllers.go:
```diff
@@ -10,17 +10,50 @@ import (
+		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+	ctx.String(200, string(stackBytes))
```
```suggestion
	ctx.String(http.StatusOK, string(stackBytes))
```

sanzoghenzo commented on swarmcd/stack.go:
```diff
@@ -22,9 +26,28 @@ type swarmStack struct {
+			log.Error("Cannot access template folder due to permission", "err", err, "folder", templateFolder)
+		}
+		templateFolder = ""
+	}
```
I realize only now that this is the wrong place to check for the directory existance.
It should be done in `updateStack` only after pulling the changes from the configured branch, to cover the edge cases in which:

- the templates folder is present only in a branch that is not the default one, or vice versa
- the templates folder has been added or removed in recent commits

Also, since the template folder is relative to the repository and not the stack itself, do you think it would be better to move some logic (folder check, templates filenames collection) to `stackRepo`?

tgaugry replied:
I think you have uncovered something greater than this.
The same thing can happen if you modify the stack definition and/or move stack files. I considered that template folder obeyed the same ruleset as the others "main" files (`config.yaml`), but that may be wrong. I'll see if this is a better fit between the call to `pullChanges` and `GenerateStack`

sanzoghenzo commented on swarmcd/init.go:
```diff
@@ -98,7 +99,7 @@ func initStacks() error {
 		}
 		discoverSecrets := config.SopsSecretsDiscovery || stackConfig.SopsSecretsDiscovery
-		swarmStack := newSwarmStack(stack, stackRepo, stackConfig.Branch, stackConfig.ComposeFile, stackConfig.SopsFiles, stackConfig.ValuesFile, discoverSecrets)
+		swarmStack := NewSwarmStack(stack, stackRepo, stackConfig.Branch, stackConfig.ComposeFile, stackConfig.SopsFiles, stackConfig.ValuesFile, discoverSecrets, config.GlobalValues, "template")
```
Apart from my comment on the NewSwarmStack function, should we make the template folder name configurable?
Should it be a stack config or a repo config?

tgaugry replied:
To me it's more a repo thing, as it is a way to group definitions, but I could see the argument of having a very big repository with say 10 stacks referring to one folder template, and the rest referring to another folder.

Let's wait until we get a little more users use case on that, for now i'll add the name config in repo definition.

sanzoghenzo commented on swarmcd/stack_test.go:
```diff
@@ -48,3 +53,58 @@ secrets:
 	}
 }
+
+// This test assume
```
complete the phrase or delete the comment

sanzoghenzo reviewed (changes requested):
Thanks again for your work.

My AI guided reviews missed a potential bug in the template folder checks.

Then I started thinking about that folder, feel free to counter argument my comments 😉

Commit:
Fixing typos/wrong inclusion

Commit:
Change templates behavior

- Add UpdateTemplatesPath method
- Move templates path setting to the repository definition
- Uniformize template(s) path/folder -> templatesPath
- Now disabled by default

tgaugry commented:
That should do the trick. It's a bit weird since the repo config cannot be changed at runtime, but this is way out of the current scope. So if you remove your templates folder, it will outoput warnings until you change `repos.yaml` configuration  and restart swarmcd. It will still work and stacks will get generated correctly (barring you actually still uses references to templates in said stacks...).

It is now also disabled by default, as I find it more natural that way instead of introducing a `templates_enabled` config variable.

tgaugry commented:
Finally found the time to deploy this after the changes from your review.
Seems like everything is working fine and every point from the review has been solved. Are we good to go with a merge ?

tgaugry committed:
Merge branch 'main' into global_values_templates

tgaugry commented:
Merged with latest changes using github webui

Commit:
Merge remote-tracking branch 'upstream/main' into global_values_templates
